src/Kernel-Tests/ProtoObjectTest.class.st

```diff
@@ -78,19 +106,38 @@ ProtoObjectTest >> testIfNotNil [
 { #category : #'tests - testing' }
 ProtoObjectTest >> testIfNotNilIfNil [
 
-	| object returnValue block |
+	| object returnValue block reached notReached|
 	object := ProtoObject new.
 	returnValue := Object new.
-	self should: [ object ifNotNil: [ self halt ] ifNil: [ self error ]  ] raise: Halt.
-	self should: [ object ifNotNil: [ :o | self halt ] ifNil: [ self error ] ] raise: Halt.
+	
+	reached := false. notReached := false.
+	object ifNotNil: [ reached := true ] ifNil: [ notReached := true ].
+	self assert: reached equals: true.
+	self assert: notReached equals: false.
+	
+	reached := false. notReached := false.
+	object ifNotNil: [ :o | reached := true ] ifNil: [ notReached := true ].
+	self assert: reached equals: true.
+	self assert: notReached equals: false.
+	
+	reached := false.
+	notReached := false.
 	self assert: (object ifNotNil: [ :o | o == object ] ifNil: [ false ]).
 	self assert: (object ifNotNil: [ returnValue ] ifNil: [ false ]) == returnValue.
 	self assert: (object ifNotNil: [ :o | returnValue ] ifNil: [ false ]) == returnValue.
 	"Now the same without inlining."
-	block := [ self halt ].
-	self should: [ object ifNotNil: block ifNil: [ self error ]  ] raise: Halt.
-	block := [ :o | self halt ].
-	self should: [ object ifNotNil: block ifNil: [ self error ] ] raise: Halt.
+	reached := false.	notReached := false.
+	block := [ reached := true ].
+	object ifNotNil: block ifNil: [ notReached := true ].
+	self assert: reached equals: true.
+	self assert: notReached equals: false.
+	
+	reached := false. notReached := false.	
+	block := [ :o | reached := true ].
+	object ifNotNil: block ifNil: [ notReached := true ].
+	self assert: reached equals: true.
+	self assert: notReached equals: false.
```

Comment:
Is really all this repeated code needed?

Reply:
It's not testing exactly the same thing each time, so... maybe?
I don't know :-)